### DISCUSSION TITLE
How to configure kustomization spec.interval with Terraform Flux Provider?

### DISCUSSION THREAD

**Discussion by zyxie1995** [fluxcd/flux2#3000] [Q&A] [answered]
Hi,

I have installed Flux v2 with [Terraform Flux Provider](https://registry.terraform.io/providers/fluxcd/flux/0.16.0).
The [interval](https://registry.terraform.io/providers/fluxcd/flux/latest/docs/data-sources/sync#interval) arg for flux_sync corresponds to GitRepository interval for checking new commits. But it's not the interval for [Kustomization](https://fluxcd.io/docs/components/kustomize/kustomization/#reconciliation) `spec.interval`(by default with Terraform Flux Provider, this Kustomization interval is 10 minutes).

In order to apply changes on the cluster faster, I would like to set Kustomization `spec.interval` to 3 minutes. Does the only solution is that I have to modify the Kustomization YAML file (apiVersion: kustomize.toolkit.fluxcd.io/v1beta2) by hand? Or is there an option offered by Terraform Flux Provider that I can configure Kustomization interval directly with Terraform?

Thanks a lot for your help and have a great day!

Zheyu

**Comment by kingdonb** [comment, marked as answer]
After bootstrapping you can change the `spec.interval` of the Flux Kustomization, in the git YAMLs.

I'm not sure if running terraform again will overwrite that change. But if you update the `flux-system` yaml through a patch, then you can make those arbitrary changes to the YAMLs that terraform provider writes into the repo.

I had a look at the `flux bootstrap` options and the limitation is the same:

```
$ flux bootstrap github --help|grep -i interval
      --interval duration       sync interval (default 1m0s)
```

There's only one sync interval parameter and it is for the GitRepository. But it may also help to know that every time GitRepository detects a change, it knows the Kustomizations which are subscribed to it. They get notified immediately (and they should reconcile themselves immediately, as long as they are not already reconciling.) So when you configure this parameter with 3m as the value for interval, you should never see a 10m delay waiting for new changes to get applied to the cluster.

I leave my Flux Kustomization interval at 10m for that reason. I do adjust the `timeout` setting, because I find the default value of (30s less than `interval`) to be a bit too long. It can happen that waiting for `timeout` makes it seem that `interval` needs to be shorter. And now I'm putting two and two together, and I've realized that setting `interval` has also the side effect of making `timeout` shorter by the same amount, down to a minimum of `30s` timeout for `1m` interval. Anyway it should solve the same issue you're experiencing to set `timeout` as short as `30s` or as long as needed, this will prevent the health checks from taking up to 9m30s (which blocks the next reconcile from beginning, the last change must either succeed or fail before another can begin.)

More: You can also set `retryInterval` so that a failure after `30s` timeout does not wait for `10m` interval, which is another minor inconvenience that you can resolve by adjusting the settings. I think these settings are not default because the behavior could be surprising, and we do want users to read the manual in order to discover these features and other adjacent ones that are unlikely to stumble on without reading the manual. These features all together would be a good blog post. Maybe show with and without Terraform, how to configure these settings. (Thanks for asking the question, it is a good one!)

You can do any of this with a patch, similar to the one for setting `interval`. So back to your original question:

To see more about patching Flux manifests with the Terraform Provider, please check out this doc: https://github.com/fluxcd/terraform-provider-flux/blob/main/docs/guides/customize-flux.md

For anyone that's reading along who didn't use Terraform, but still has exactly the same issue, it is a slightly different answer:
https://fluxcd.io/flux/installation/#customize-flux-manifests

**Reply by zyxie1995** [reply]
Thank you very much for your answer @kingdonb!

**Comment by alexfd** [comment]
Hello, can you please provide more clarity to this issue. Where do you mean to update the yaml? In gotk-sync.yaml? It clearly says not to edit that file.
The problem I am facing, similar to @zyxie1995, is when I bootstrap the git repo flux automatically creates the gotk-sync.yaml file with a Kustomization of 10 minute interval. How do I update this?

**Reply by kingdonb** [reply]
It's intended that you will do this in the `kustomization.yaml` – if you need to patch Flux resources in one of the files marked "DO NOT EDIT" and you've stopped, and not edited them in place, 🙏 👍 great job paying attention, most people ignore this warning even though it's in all caps :D (Oh no)

You can follow this example (where I am patching deployments only, but you can patch any resource type, including Flux resources like GitRepository and Kustomization):

https://github.com/kingdon-ci/fleet-infra/blob/main/clusters/test/limnocentral/flux-system/kustomization.yaml

None of the examples I found patch GitRepository directly, but there is one here that patches a Flux Kustomization:
https://fluxcd.io/flux/cheatsheets/bootstrap/#multi-tenancy-lockdown

I [just updated](https://github.com/kingdon-ci/fleet-infra/blob/b98d7a1eba4002b62b5faab18478d920e8f4370c/clusters/test/limnocentral/flux-system/kustomization.yaml#L64-L73) my Kustomization linked above to add a patch for setting kustomization.spec.interval to a different value, just to prove that it would work. And it does work, but if you are using Terraform provider like I have done here instead of normal Flux bootstrap, you need to be aware that it can overwrite `kustomization.yaml` (which is I think a divergence in behavior compared to regular Flux CLI, but probably more in line with what Terraform users were likely to expect – any customizations should have been reflected in your `main.tf` where Flux bootstrap is done, and not bolted on separately after the fact! But if you're not using Terraform, ignore this...)

There is more Flux documentation that supports this idea of customizing your Flux install here:
https://fluxcd.io/flux/installation/#customize-flux-manifests

It's further reinforced all throughout the Bootstrap cheatsheet and other docs related to customizing Flux installs, which reiterate the same information but for specific audiences, and with more examples:
https://fluxcd.io/flux/cheatsheets/bootstrap/

**Reply by alexfd** [reply]
hah, thanks @kingdonb. After posting here I also came across patching so I assumed this was the correct way, but haven't managed to make it work. Only way I fixed it was to run `flux create kustomization --interval=1m`.  I'll give your instructions a try, thank you.
